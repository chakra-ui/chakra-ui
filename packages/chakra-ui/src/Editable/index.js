/** @jsx jsx */
import { jsx } from "@emotion/core";
import {
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
  forwardRef,
} from "react";
import PseudoBox from "../PseudoBox";
import Box from "../Box";

const EditableContext = createContext();

const Editable = forwardRef(
  (
    {
      value: valueProp,
      defaultValue,
      isDisabled,
      onChange,
      startWithEditView,
      onCancel,
      onSubmit,
      selectAllOnFocus = true,
      submitOnBlur = true,
      isPreviewFocusable = true,
      placeholder = "Click to edit...",
      children,
      onEdit,
      ...rest
    },
    ref,
  ) => {
    const [isEditing, setIsEditing] = useState(
      startWithEditView && !isDisabled,
    );
    const { current: isControlled } = useRef(valueProp != null);

    const [value, setValue] = useState(defaultValue || "");
    const _value = isControlled ? valueProp : value;
    const [previousValue, setPreviousValue] = useState(_value);

    const inputRef = useRef(null);

    const onRequestEdit = event => {
      if (!isDisabled) {
        setIsEditing(true);
      }
    };

    useEffect(() => {
      if (isEditing) {
        onEdit && onEdit();
      }
    }, [isEditing, onEdit]);

    useEffect(() => {
      if (isEditing && inputRef.current) {
        inputRef.current.focus();
        selectAllOnFocus && inputRef.current.select();
      }
    }, [isEditing, selectAllOnFocus]);

    const handleCancel = () => {
      setIsEditing(false);
      setValue(previousValue);
      if (value !== previousValue) {
        onChange && onChange(previousValue);
      }
      onCancel && onCancel(previousValue);
    };

    const handleSubmit = () => {
      setIsEditing(false);
      setPreviousValue(value);
      onSubmit && onSubmit(value);
    };

    const handleChange = event => {
      const { value } = event.target;
      if (!isControlled) {
        setValue(value);
      }
      onChange && onChange(value);
    };

    const handleKeyDown = event => {
      const { key } = event;
      if (key === "Escape") {
        handleCancel();
        return;
      }

      if (key === "Enter" && !event.shiftKey && !event.metaKey) {
        handleSubmit();
      }
    };

    const handleFocus = event => {
      if (selectAllOnFocus) {
        inputRef.current.select();
      }
    };

    const childContext = {
      inputRef,
      isEditing,
      isDisabled,
      placeholder,
      onRequestEdit,
      submitOnBlur,
      isPreviewFocusable,
      value: _value,
      onKeyDown: handleKeyDown,
      onChange: handleChange,
      onSubmit: handleSubmit,
      onCancel: handleCancel,
      onFocus: handleFocus,
    };

    return (
      <EditableContext.Provider value={childContext}>
        <Box ref={ref} {...rest}>
          {typeof children === "function"
            ? children({
                isEditing,
                onSubmit: handleSubmit,
                onCancel: handleCancel,
                onRequestEdit,
              })
            : children}
        </Box>
      </EditableContext.Provider>
    );
  },
);

Editable.displayName = "Editable";

const sharedProps = {
  fontSize: "inherit",
  fontWeight: "inherit",
  textAlign: "inherit",
  bg: "transparent",
  transition: "all 0.2s",
  borderRadius: "md",
  px: "3px",
  mx: "-3px",
};

export const EditablePreview = props => {
  const {
    isEditing,
    isDisabled,
    value,
    onRequestEdit,
    placeholder,
    isPreviewFocusable,
  } = useContext(EditableContext);
  const hasValue = value != null && value !== "";

  const getTabIndex = () => {
    if ((!isEditing || !isDisabled) && isPreviewFocusable) {
      return 0;
    }

    return null;
  };

  const styleProps = {
    ...sharedProps,
    cursor: "text",
    display: "inline-block",
    opacity: !hasValue ? 0.6 : undefined,
  };

  if (isEditing) {
    return null;
  }

  return (
    <PseudoBox
      as="span"
      aria-disabled={isDisabled}
      tabIndex={getTabIndex()}
      onFocus={onRequestEdit}
      {...styleProps}
      {...props}
    >
      {hasValue ? value : placeholder}
    </PseudoBox>
  );
};

export const EditableInput = props => {
  const {
    inputRef,
    isEditing,
    onChange,
    onKeyDown,
    value,
    onSubmit,
    onCancel,
    submitOnBlur,
    placeholder,
    isDisabled,
  } = useContext(EditableContext);

  if (!isEditing) {
    return null;
  }

  const styleProps = {
    ...sharedProps,
    width: "full",
    _placeholder: {
      opacity: "0.6",
    },
  };

  const renderProps = {
    ref: inputRef,
    onBlur: event => {
      submitOnBlur ? onSubmit() : onCancel();
      if (props.onBlur) {
        props.onBlur(event);
      }
    },
    value,
    placeholder,
    onChange,
    onKeyDown,
  };

  return props.children ? (
    props.children(renderProps)
  ) : (
    <PseudoBox
      aria-disabled={isDisabled}
      disabled={isDisabled}
      as="input"
      outline="none"
      _focus={{ shadow: "outline" }}
      {...renderProps}
      {...styleProps}
      {...props}
    />
  );
};

export default Editable;
