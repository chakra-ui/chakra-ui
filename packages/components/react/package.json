{
  "name": "@chakra-ui/react",
  "version": "2.4.9",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "src/index.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@chakra-ui/accordion": "workspace:*",
    "@chakra-ui/alert": "workspace:*",
    "@chakra-ui/avatar": "workspace:*",
    "@chakra-ui/breadcrumb": "workspace:*",
    "@chakra-ui/button": "workspace:*",
    "@chakra-ui/card": "workspace:*",
    "@chakra-ui/checkbox": "workspace:*",
    "@chakra-ui/close-button": "workspace:*",
    "@chakra-ui/control-box": "workspace:*",
    "@chakra-ui/counter": "workspace:*",
    "@chakra-ui/css-reset": "workspace:*",
    "@chakra-ui/editable": "workspace:*",
    "@chakra-ui/form-control": "workspace:*",
    "@chakra-ui/hooks": "workspace:*",
    "@chakra-ui/icon": "workspace:*",
    "@chakra-ui/image": "workspace:*",
    "@chakra-ui/input": "workspace:*",
    "@chakra-ui/layout": "workspace:*",
    "@chakra-ui/live-region": "workspace:*",
    "@chakra-ui/media-query": "workspace:*",
    "@chakra-ui/menu": "workspace:*",
    "@chakra-ui/modal": "workspace:*",
    "@chakra-ui/number-input": "workspace:*",
    "@chakra-ui/pin-input": "workspace:*",
    "@chakra-ui/popover": "workspace:*",
    "@chakra-ui/popper": "workspace:*",
    "@chakra-ui/portal": "workspace:*",
    "@chakra-ui/progress": "workspace:*",
    "@chakra-ui/provider": "workspace:*",
    "@chakra-ui/radio": "workspace:*",
    "@chakra-ui/react-env": "workspace:*",
    "@chakra-ui/select": "workspace:*",
    "@chakra-ui/skeleton": "workspace:*",
    "@chakra-ui/slider": "workspace:*",
    "@chakra-ui/spinner": "workspace:*",
    "@chakra-ui/stat": "workspace:*",
    "@chakra-ui/styled-system": "workspace:*",
    "@chakra-ui/switch": "workspace:*",
    "@chakra-ui/system": "workspace:*",
    "@chakra-ui/table": "workspace:*",
    "@chakra-ui/tabs": "workspace:*",
    "@chakra-ui/tag": "workspace:*",
    "@chakra-ui/textarea": "workspace:*",
    "@chakra-ui/theme": "workspace:*",
    "@chakra-ui/theme-utils": "workspace:*",
    "@chakra-ui/toast": "workspace:*",
    "@chakra-ui/tooltip": "workspace:*",
    "@chakra-ui/transition": "workspace:*",
    "@chakra-ui/utils": "workspace:*",
    "@chakra-ui/visually-hidden": "workspace:*"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": ">=4.0.0",
    "react": ">=18",
    "react-dom": ">=18"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/components/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://storybook.chakra-ui.com"
  },
  "devDependencies": {
    "@emotion/react": "^11.9.0",
    "@emotion/styled": "^11.8.1",
    "framer-motion": "^6.2.9",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "clean-package": "2.1.1"
  },
  "scripts": {
    "build": "tsup src --dts",
    "dev": "pnpm build:fast -- --watch",
    "clean": "rimraf dist .turbo",
    "typecheck": "tsc --noEmit",
    "build:fast": "tsup src",
    "prepack": "clean-package",
    "postpack": "clean-package restore"
  },
  "clean-package": "../../../clean-package.config.json",
  "tsup": {
    "clean": true,
    "target": "es2019",
    "format": [
      "cjs",
      "esm"
    ]
  }
}
