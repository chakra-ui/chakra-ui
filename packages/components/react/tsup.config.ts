import fs from "node:fs"
import path from "node:path"
import { defineConfig } from "tsup"

function removeDirective(fileName: string) {
  const file = path.resolve(__dirname, "dist", fileName)
  return fs.promises.writeFile(
    file,
    fs.readFileSync(file, "utf-8").replace("'use client'\n", ""),
  )
}

export default defineConfig({
  format: ["cjs", "esm"],
  target: "es2019",
  sourcemap: true,
  banner: {
    js: "'use client'",
  },
  async onSuccess() {
    await Promise.all([
      removeDirective("index.mjs"),
      removeDirective("index.js"),
    ])
  },
})
