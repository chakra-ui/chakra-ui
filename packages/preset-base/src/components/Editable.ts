import { ComponentTheme } from "./utils"
import { SystemProps } from "@chakra-ui/system"

const sharedProps: SystemProps = {
  fontSize: "inherit",
  fontWeight: "inherit",
  textAlign: "inherit",
  bg: "transparent",
  transition: "all 0.2s",
  borderRadius: "md",
  paddingX: "3px",
  marginX: "-3px",
}

const Editable: ComponentTheme = {
  baseStyle: {
    Preview: {
      ...sharedProps,
      cursor: "text",
      display: "inline-block",
    },
    Input: {
      ...sharedProps,
      outline: 0,
      width: "full",
      _focus: {
        boxShadow: "outline",
      },
      _placeholder: {
        opacity: 0.6,
      },
    },
  },
}

export default Editable
