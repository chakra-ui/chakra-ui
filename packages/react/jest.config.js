const baseConfig = require("../../jest.config")

module.exports = {
  ...baseConfig,
}
