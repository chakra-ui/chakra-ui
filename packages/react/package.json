{
  "name": "@chakra-ui/react",
  "version": "1.6.12",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/types/index.d.ts",
  "typings": "dist/types/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/cjs/index.js",
      "default": "./dist/esm/index.js"
    }
  },
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "start": "nodemon --watch src --exec yarn build -e ts,tsx",
    "build": "concurrently yarn:build:*",
    "test": "jest --env=jsdom --passWithNoTests",
    "lint": "concurrently yarn:lint:*",
    "version": "yarn build",
    "build:esm": "cross-env BABEL_ENV=esm babel src --root-mode upward --extensions .ts,.tsx -d dist/esm --source-maps",
    "build:cjs": "cross-env BABEL_ENV=cjs babel src --root-mode upward --extensions .ts,.tsx -d dist/cjs --source-maps",
    "build:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types",
    "test:cov": "yarn test --coverage",
    "lint:src": "eslint src --ext .ts,.tsx --config ../../.eslintrc",
    "lint:types": "tsc --noEmit"
  },
  "dependencies": {
    "@chakra-ui/accordion": "1.3.10",
    "@chakra-ui/alert": "1.2.9",
    "@chakra-ui/avatar": "1.2.12",
    "@chakra-ui/breadcrumb": "1.2.10",
    "@chakra-ui/button": "1.4.6",
    "@chakra-ui/checkbox": "1.5.10",
    "@chakra-ui/close-button": "1.1.13",
    "@chakra-ui/control-box": "1.0.17",
    "@chakra-ui/counter": "1.1.11",
    "@chakra-ui/css-reset": "1.0.0",
    "@chakra-ui/editable": "1.2.11",
    "@chakra-ui/form-control": "1.4.3",
    "@chakra-ui/hooks": "1.6.2",
    "@chakra-ui/icon": "1.1.13",
    "@chakra-ui/image": "1.0.22",
    "@chakra-ui/input": "1.2.12",
    "@chakra-ui/layout": "1.4.11",
    "@chakra-ui/live-region": "1.0.16",
    "@chakra-ui/media-query": "1.1.5",
    "@chakra-ui/menu": "1.7.8",
    "@chakra-ui/modal": "1.9.4",
    "@chakra-ui/number-input": "1.2.12",
    "@chakra-ui/pin-input": "1.6.7",
    "@chakra-ui/popover": "1.9.1",
    "@chakra-ui/popper": "2.3.1",
    "@chakra-ui/portal": "1.2.11",
    "@chakra-ui/progress": "1.1.16",
    "@chakra-ui/provider": "1.6.11",
    "@chakra-ui/radio": "1.3.13",
    "@chakra-ui/react-env": "1.0.8",
    "@chakra-ui/select": "1.1.17",
    "@chakra-ui/skeleton": "1.1.21",
    "@chakra-ui/slider": "1.4.2",
    "@chakra-ui/spinner": "1.1.14",
    "@chakra-ui/stat": "1.1.14",
    "@chakra-ui/switch": "1.2.13",
    "@chakra-ui/system": "1.7.6",
    "@chakra-ui/table": "1.2.8",
    "@chakra-ui/tabs": "1.5.7",
    "@chakra-ui/tag": "1.1.14",
    "@chakra-ui/textarea": "1.1.16",
    "@chakra-ui/theme": "1.11.1",
    "@chakra-ui/toast": "1.3.4",
    "@chakra-ui/tooltip": "1.3.14",
    "@chakra-ui/transition": "1.3.8",
    "@chakra-ui/utils": "1.8.4",
    "@chakra-ui/visually-hidden": "1.0.16"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": "3.x || 4.x || 5.x",
    "react": ">=16.8.6",
    "react-dom": ">=16.8.6"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://chakra-ui.netlify.app"
  },
  "devDependencies": {
    "@emotion/react": "^11.1.4",
    "@emotion/styled": "^11.0.0",
    "framer-motion": "^4.0.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.1"
  }
}
