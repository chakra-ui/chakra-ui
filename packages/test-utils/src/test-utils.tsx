import { ChakraProvider } from "@chakra-ui/react"
import "@testing-library/jest-dom/extend-expect"
import {
  render as rtlRender,
  RenderOptions,
  fireEvent,
  RenderResult,
} from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations, axe } from "jest-axe"
import { createSerializer } from "@emotion/jest"
import { RunOptions } from "axe-core"

expect.addSnapshotSerializer(createSerializer())
expect.extend(toHaveNoViolations)

type UI = Parameters<typeof rtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, "wrapper"> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult =>
  rtlRender(
    <ChakraProvider>
      <Wrapper>{ui}</Wrapper>
    </ChakraProvider>,
    options,
  )

export { rtlRender }
export { axe }

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })

type TestA11YOptions = TestOptions & { axeOptions?: RunOptions }

/**
 * Validates against common a11y mistakes.
 *
 * Wrapper for jest-axe
 *
 * @example
 * ```jsx
 * it('passes a11y test', async () => {
 *  await testA11Y(<MyComponent />, options);
 * });
 *
 * // sometimes we need to perform interactions first to render conditional UI
 * it('passes a11y test when open', async () => {
 *  const { container } = render(<MyComponent />, options);
 *
 *  fireEvent.click(screen.getByRole('button'));
 *
 *  await testA11Y(container, options);
 * });
 * ```
 *
 * @see https://github.com/nickcolley/jest-axe#testing-react-with-react-testing-library
 */
export const testA11y = async (
  ui: UI | Element,
  { axeOptions, ...options }: TestA11YOptions = {},
) => {
  const container = React.isValidElement(ui)
    ? render(ui, options).container
    : ui

  const results = await axe(container, axeOptions)

  expect(results).toHaveNoViolations()
}
