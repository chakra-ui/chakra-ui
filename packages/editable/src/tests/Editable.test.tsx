import * as React from "react"
import { render, userEvent, fireEvent } from "@chakra-ui/test-utils"
import { Editable, EditableInput, EditablePreview } from "../Editable"

test("should match snapshot", () => {
  const utils = render(
    <Editable defaultValue="testing">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )

  expect(utils.asFragment()).toMatchSnapshot()
})

test("should render properly", () => {
  const utils = render(
    <Editable defaultValue="testing">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )

  const preview = utils.getByTestId("preview")
  const input = utils.getByTestId("input")

  expect(input).toHaveAttribute("hidden")
  expect(preview).toHaveTextContent("testing")
})

test("uncontrolled: handles callbacks correctly", async () => {
  const onChange = jest.fn()
  const onCancel = jest.fn()
  const onSubmit = jest.fn()
  const onEdit = jest.fn()

  const utils = render(
    <Editable
      onChange={onChange}
      onCancel={onCancel}
      onSubmit={onSubmit}
      onEdit={onEdit}
      defaultValue="Hello "
    >
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )
  const preview = utils.getByTestId("preview")
  const input = utils.getByTestId("input")

  // calls `onEdit` when preview is focused
  fireEvent.focus(preview)
  expect(onEdit).toHaveBeenCalled()

  // calls `onChange` with input on change
  userEvent.type(input, "World")
  expect(onChange).toHaveBeenCalledWith("Hello World")

  // calls `onCancel` with previous value when "esc" pressed
  fireEvent.keyDown(input, { key: "Escape" })
  expect(onCancel).toHaveBeenCalledWith("Hello ")

  // calls `onSubmit` with previous value when "enter" pressed after cancelling
  fireEvent.keyDown(input, { key: "Enter" })
  expect(onSubmit).toHaveBeenCalledWith("Hello ")

  // returns new value when submitting without cancelling
  userEvent.type(input, "World")
  fireEvent.keyDown(input, { key: "Enter" })
  expect(onSubmit).toHaveBeenCalledWith("Hello World")

  // cancelling now returns new value
  fireEvent.keyDown(input, { key: "Escape" })
  expect(onCancel).toHaveBeenCalledWith("Hello World")
})

test("controlled: handles callbacks correctly", () => {
  const onChange = jest.fn()
  const onCancel = jest.fn()
  const onSubmit = jest.fn()
  const onEdit = jest.fn()

  const utils = render(
    <Editable
      onChange={onChange}
      onCancel={onCancel}
      onSubmit={onSubmit}
      onEdit={onEdit}
      value="Hello "
    >
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )
  const preview = utils.getByTestId("preview")
  const input = utils.getByTestId("input")

  // calls `onEdit` when preview is focused
  fireEvent.focus(preview)
  expect(onEdit).toHaveBeenCalled()

  // calls `onChange` with new input on change
  userEvent.type(input, "World")
  expect(onChange).toHaveBeenCalledWith("Hello World")

  // calls `onSubmit` with `value`
  fireEvent.keyDown(input, { key: "Enter" })
  expect(onSubmit).toHaveBeenCalledWith("Hello ")

  // calls `onCancel` with `value`
  fireEvent.keyDown(input, { key: "Escape" })
  expect(onSubmit).toHaveBeenCalledWith("Hello ")
})

test("handles preview and input callbacks", () => {
  const onFocus = jest.fn()
  const onBlur = jest.fn()
  const onChange = jest.fn()
  const onKeyDown = jest.fn()

  const utils = render(
    <Editable defaultValue="Hello ">
      <EditablePreview onFocus={onFocus} data-testid="preview" />
      <EditableInput
        onBlur={onBlur}
        onChange={onChange}
        onKeyDown={onKeyDown}
        data-testid="input"
      />
    </Editable>,
  )
  const preview = utils.getByTestId("preview")
  const input = utils.getByTestId("input")

  // calls `onFocus` when preview is focused
  fireEvent.focus(preview)
  expect(onFocus).toHaveBeenCalled()

  // calls `onChange` when input is changed
  userEvent.type(input, "World")
  expect(onChange).toHaveBeenCalled()

  // calls `onKeyDown` when key is pressed in input
  fireEvent.keyDown(input, { key: "Escape" })
  expect(onKeyDown).toHaveBeenCalled()

  // calls `onBlur` when focus leaves input
  fireEvent.focus(preview)
  expect(onBlur).toHaveBeenCalled()
})

test("has the proper aria attributes", () => {
  const utils = render(
    <Editable defaultValue="">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )
  const preview = utils.getByTestId("preview")
  const input = utils.getByTestId("input")

  // preview and input do not have aria-disabled when `Editable` is not disabled
  expect(preview).not.toHaveAttribute("aria-disabled")
  expect(input).not.toHaveAttribute("aria-disabled")

  utils.rerender(
    <Editable isDisabled defaultValue="">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )

  // preview and input have aria-disabled when `Editable` is disabled
  expect(preview).toHaveAttribute("aria-disabled", "true")
  expect(input).toHaveAttribute("aria-disabled", "true")
})

test("can submit on blur", () => {
  const onSubmit = jest.fn()
  const utils = render(
    <Editable submitOnBlur onSubmit={onSubmit} defaultValue="testing">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )
  const input = utils.getByTestId("input")

  fireEvent.blur(input)
  expect(onSubmit).toHaveBeenCalledWith("testing")
})
