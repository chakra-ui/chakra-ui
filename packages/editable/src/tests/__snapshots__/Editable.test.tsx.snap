// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-chakra-editable=""
  >
    <span
      data-chakra-editable-preview=""
      data-testid="preview"
    >
      testing
    </span>
    <input
      data-chakra-editable-input=""
      data-testid="input"
      hidden=""
      value="testing"
    />
  </div>
</DocumentFragment>
`;
